PR: Halaman State Management (#73)
vuejs-id/docs: src/v2/guide/state-management.md

---
title: Pengelolaan State
type: guide
order: 502
---

> ## Implementasi Seperti Flux Resmi

Reviewer: Saya rasa Flux like tidak perlu diterjemahkan tapi diberi penjelasan
```suggestion
## Implementasi Flux-Like (Sejenis Flux) Resmi
```

> Aplikasi besar bisa tumbuh menjadi sangat kompleks, dikarnakan banyak sekali bagian dari state yang tersebar ke components dan berinteraksi dengan mereka. Untuk mengatasi masalah ini, Vue menawarkan [vuex](https://github.com/vuejs/vuex): Library pengelolaan state milik kami sendiri. Ia bahkan terintegrasi ke [vue-devtools](https://github.com/vuejs/vue-devtools), dan juga tidak perlu mempersiapkan apa-apa untuk mengakses [time travel debugging](https://raw.githubusercontent.com/vuejs/vue-devtools/master/media/demo.gif).

Reviewer 2: ```suggestion
Aplikasi besar bisa tumbuh menjadi sangat kompleks, dikarenakan banyak sekali bagian dari state yang tersebar ke seluruh komponen dan berinteraksi dengan mereka. Untuk mengatasi masalah ini, Vue menawarkan [vuex](https://github.com/vuejs/vuex): Sebuah pustaka yang berfungsi untuk mengelola state buatan kami sendiri. Ia bahkan mampu terintegrasi dengan [vue-devtools](https://github.com/vuejs/vue-devtools), dan juga kita tidak perlu mempersiapkan apa-apa (tidak perlu melakukan konfigurasi khusus) untuk bisa mengakses [time travel debugging](https://raw.githubusercontent.com/vuejs/vue-devtools/master/media/demo.gif).
```
Masih kurang natural, sudah saya perbaiki seperti diatas.

> <div class="vue-mastery"><a href="https://www.vuemastery.com/courses/mastering-vuex/intro-to-vuex/" target="_blank" rel="noopener" title="Vuex Tutorial">Lihatlah sebuah video yang menjelaskan tentang Vue Mastery</a></div>

Reviewer 2: ```suggestion
<div class="vue-mastery"><a href="https://www.vuemastery.com/courses/mastering-vuex/intro-to-vuex/" target="_blank" rel="noopener" title="Vuex Tutorial">Tonton video penjelasannya di Vue Mastery</a></div>
```

> <div class="vue-mastery"><a href="https://www.vuemastery.com/courses/mastering-vuex/intro-to-vuex/" target="_blank" rel="noopener" title="Vuex Tutorial">Lihatlah sebuah video yang menjelaskan tentang Vue Mastery</a></div>

Reviewer 3: Tonton video penjelasan di Vue Mastery

> <div class="vue-mastery"><a href="https://www.vuemastery.com/courses/mastering-vuex/intro-to-vuex/" target="_blank" rel="noopener" title="Vuex Tutorial">Lihatlah sebuah video yang menjelaskan tentang Vue Mastery</a></div>

Reviewer 2: ```suggestion
<div class="vue-mastery"><a href="https://www.vuemastery.com/courses/mastering-vuex/intro-to-vuex/" target="_blank" rel="noopener" title="Vuex Tutorial">Tonton video penjelasannya di Vue Mastery</a></div>
```
natural

### Informasi untuk Developer React

> Jika kamu pernah menggunakan React, kamu mungkin ingin tahu perbandingan vuex dan [redux](https://github.com/reactjs/redux), Libary untuk implementasi Flux yang paling populer. Redux sebenarnya hanya sebatas lapisan tampilan saja, jadi bisa digunakan bersama Vue dengan mudah dengan cara via [simple bindings](https://yarnpkg.com/en/packages?q=redux%20vue&p=1). Vuex berbeda karna ia langsung terhubung dengan Aplikasi milik Vue. Jadi karna ini ia terintegrasi lebih baik dengan Vue, dan juga Vuex menawarkan API yang lebih intuitif dan meningkatkan pengalaman pengembangan.

Reviewer 3: Library -> Pustaka

Reviewer 3: Redux sebenarnya tidak tergantung pada suatu lapisan tampilan

Reviewer 3: dengan cara via -> dengan cara

Reviewer 3: simple binding -> binding sederhana

> Jika kamu pernah menggunakan React, kamu mungkin ingin tahu perbandingan vuex dan [redux](https://github.com/reactjs/redux), Libary untuk implementasi Flux yang paling populer. Redux sebenarnya hanya sebatas lapisan tampilan saja, jadi bisa digunakan bersama Vue dengan mudah dengan cara via [simple bindings](https://yarnpkg.com/en/packages?q=redux%20vue&p=1). Vuex berbeda karna ia langsung terhubung dengan Aplikasi milik Vue. Jadi karna ini ia terintegrasi lebih baik dengan Vue, dan juga Vuex menawarkan API yang lebih intuitif dan meningkatkan pengalaman pengembangan.

Reviewer 2: ```suggestion
Jika Anda pernah menggunakan React, Anda mungkin ingin tahu perbandingan antara vuex dan [redux](https://github.com/reactjs/redux), sebuah pustaka untuk implementasi Flux yang paling populer. Redux sebenarnya hanya sebatas lapisan di tampilan saja, jadi bisa digunakan bersama-sama dengan Vue secara mudah melalui [simple bindings](https://yarnpkg.com/en/packages?q=redux%20vue&p=1). Vuex berbeda karena ia terhubung dengan aplikasi Vue secara langsung. Oleh sebab itu, Vuex ini bisa lebih baik diintegrasikan dengan Vue, dan juga Vuex menawarkan API yang lebih intuitif serta meningkatkan pengalaman para pengembang.
```
Ada beberapa kata yang harus diubah agar lebih enak dibaca
seperti pengalaman pengembangan, boleh kita ubah jadi pengalaman para pengembang.

> Jika kamu pernah menggunakan React, kamu mungkin ingin tahu perbandingan vuex dan [redux](https://github.com/reactjs/redux), Libary untuk implementasi Flux yang paling populer. Redux sebenarnya hanya sebatas lapisan tampilan saja, jadi bisa digunakan bersama Vue dengan mudah dengan cara via [simple bindings](https://yarnpkg.com/en/packages?q=redux%20vue&p=1). Vuex berbeda karna ia langsung terhubung dengan Aplikasi milik Vue. Jadi karna ini ia terintegrasi lebih baik dengan Vue, dan juga Vuex menawarkan API yang lebih intuitif dan meningkatkan pengalaman pengembangan.

Reviewer 2: ```suggestion
Jika Anda pernah menggunakan React, Anda mungkin ingin tahu perbandingan antara vuex dengan [redux](https://github.com/reactjs/redux), sebuah pustaka untuk implementasi Flux yang paling populer. Redux sebenarnya hanya sebatas lapisan di tampilan saja, jadi bisa digunakan bersama - sama dengan Vue secara mudah melalui [simple bindings](https://yarnpkg.com/en/packages?q=redux%20vue&p=1). Vuex berbeda karena ia langsung terhubung dengan aplikasi Vue. Oleh karena itu, Vuex lebih cocok jika di integrasikan dengan Vue, dan juga Vuex menawarkan API yang lebih intuitif dan mampu meningkatkan pengalaman para pengembang.
```
- ada typo
- kamu => Anda
- kurang natural

pengalaman pengembangan => terdengar sangat aneh, maka kita boleh ubah menjadi
pengalaman para pengembang. Maksud berbeda, tapi tujuan yang di sampaikan sama. Jadi boleh untuk merubah maksud asalkan tujuan nya sama. Biar lebih natural dan enak dibaca

> ## Pengelolaan State dari Awal Dengan Mudah

Reviewer 3: Pengelolaan State Sederhana dari Awal

> Jika anda pernah menggunakan React, anda mungkin ingin tahu perbandingan vuex dan [redux](https://github.com/reactjs/redux), Pustaka untuk implementasi Flux yang paling populer. Redux sebenarnya tidak tergantung pada suatu lapisan tampilan, jadi bisa digunakan bersama Vue dengan mudah dengan cara [binding sederhana](https://yarnpkg.com/en/packages?q=redux%20vue&p=1). Vuex berbeda karna ia langsung terhubung dengan Aplikasi milik Vue. Jadi karna ini ia terintegrasi lebih baik dengan Vue, dan juga Vuex menawarkan API yang lebih intuitif dan sekaligus meningkatkan pengalaman dan kenyamanan pengembangan.

Reviewer 2: ```suggestion
Jika Anda pernah menggunakan React, Anda mungkin ingin tahu perbandingan antara vuex dengan [redux](https://github.com/reactjs/redux), sebuah pustaka untuk implementasi Flux yang paling populer. Redux sebenarnya hanya sebatas lapisan di tampilan saja, jadi bisa digunakan bersama - sama dengan Vue secara mudah melalui [simple bindings](https://yarnpkg.com/en/packages?q=redux%20vue&p=1). Vuex berbeda karena ia langsung terhubung dengan aplikasi Vue. Oleh karena itu, Vuex lebih cocok jika di integrasikan dengan Vue, dan juga Vuex menawarkan API yang lebih intuitif dan mampu meningkatkan pengalaman para pengembang.
```

> Sering diabaikan bahwa sumber kebenaran di aplikasi Vue adalah `data` objek yang mentah - sebuah instance Vue hanya wakil untuk mengakasesnya. Karna itu, jika kamu memiliki bagian state yang ingin di bagikan pada beberapa instance, kamu bisa membaginya dengan identitas:

Reviewer 2: ```suggestion
Hal ini seringkali diabaikan bahwa sumber kebenaran di aplikasi Vue adalah objek `data` yang mentah - hanya *instance* Vue yang boleh mengakasesnya. Oleh karena itu, jika Anda memiliki beberapa state yang ingin di bagikan ke beberapa *instance*, Anda bisa membaginya dengan identitas:
```
kamu => Anda
instance pakai italic
kurang natural

> Sering diabaikan bahwa sumber kebenaran di aplikasi Vue adalah `data` objek yang mentah - sebuah instance Vue hanya wakil untuk mengakasesnya. Karna itu, jika kamu memiliki bagian state yang ingin di bagikan pada beberapa instance, kamu bisa membaginya dengan identitas:

Reviewer 2: ```suggestion
Hal ini sering diabaikan bahwa sumber kebenaran di aplikasi Vue adalah objek `data` yang mentah - hanya instance Vue saja yang boleh mengaksesnya. Oleh karna itu, jika Anda memiliki beberapa *state* yang ingin di bagikan ke beberapa instance yang lain, Anda bisa membaginya dengan identitas:
```
kurang natural dan ada typo

> Hal ini seringkali diabaikan bahwa sumber kebenaran di aplikasi Vue adalah objek `data` yang mentah - hanya **instance** Vue yang boleh mengakasesnya. Oleh karena itu, jika Anda memiliki beberapa state yang ingin di bagikan ke beberapa **instance**, Anda bisa membaginya dengan identitas:

Reviewer 2: instance pakai single *, bukan double **
kalau double ** jadi nya bold, bukan italic

mengakasesnya => mengaksesnya

})
```

> Sekarang saat `Sumber Kebenaran` dirubah, kedua `vmA` dan `vmB` akan merubah tampilan mereka secara otomatis. Subcomponents dalam beberapa instance ini juga perlu mengaksesnya dengan `this.$root.$data`. Sekarang kita punya satu sumber kebenaran, namun melakukan debugging akan sangat sulit sekali. Setiap bagian dari data bisa berubah dengan setiap bagian dari aplikasi kita setiap waktunya, tanpa meninggalkan bekas.

Reviewer 2: ```suggestion
Sekarang kapanpun `Sumber Kebenaran` diubah, kedua `vmA` dan `vmB` akan merubah tampilan mereka secara otomatis. Sub komponen dalam beberapa *instance* ini juga bisa mengaksesnya dengan cara `this.$root.$data`. Sekarang kita punya satu sumber kebenaran, namun ketika melakukan proses *debugging* akan terasa sangat sulit sekali. Setiap bagian dari data bisa berubah dengan setiap bagian dari aplikasi kita setiap waktunya, tanpa meninggalkan bekas sama sekali.
```
naturalisasi

> Sekarang saat `Sumber Kebenaran` dirubah, kedua `vmA` dan `vmB` akan merubah tampilan mereka secara otomatis. Subcomponents dalam beberapa instance ini juga perlu mengaksesnya dengan `this.$root.$data`. Sekarang kita punya satu sumber kebenaran, namun melakukan debugging akan sangat sulit sekali. Setiap bagian dari data bisa berubah dengan setiap bagian dari aplikasi kita setiap waktunya, tanpa meninggalkan bekas.

Reviewer 3: sourceOfTruth gak boleh diterjemahkan disini

> Sekarang saat `Sumber Kebenaran` dirubah, kedua `vmA` dan `vmB` akan merubah tampilan mereka secara otomatis. Subcomponents dalam beberapa instance ini juga perlu mengaksesnya dengan `this.$root.$data`. Sekarang kita punya satu sumber kebenaran, namun melakukan debugging akan sangat sulit sekali. Setiap bagian dari data bisa berubah dengan setiap bagian dari aplikasi kita setiap waktunya, tanpa meninggalkan bekas.

Reviewer 2: ```suggestion
Sekarang, kapanpun `Sumber Kebenaran` diubah, kedua `vmA` dan `vmB` akan merubah tampilan mereka secara otomatis. Sub komponen yang ada di dalam beberapa *instance* ini juga bisa mengaksesnya dengan cara `this.$root.$data`. Sekarang kita sudah punya satu sumber kebenaran, namun untuk melakukan proses *debugging* akan terasa sangat sulit sekali. Setiap bagian dari data bisa berubah dikarenakan perubahan dari setiap bagian dari aplikasi kita, setiap waktunya, tanpa meninggalkan bekas sama sekali.
```
kurang natural

}
```

> Perhatikan semua action yang merubah state dari store diletakkan didalam store itu sendiri. Tipe pengelolaan state yang terpusat ini membuat semakin mudah untuk mengerti bahwa tipe perubahan apa yang bisa terjadi dan bagaimana mereka dijalankan. Sekarang saat terjadi kesalahan, kita juga memiliki sebuah catatan kenapa bisa terjadi bug.

Reviewer 2: ```suggestion
Perhatikan semua *action* yang merubah *state* dari *store* diletakkan didalam *store* itu sendiri. Tipe pengelolaan state yang terpusat seperti ini semakin memudahkan kita untuk mengerti bahwa tipe perubahan apa yang bisa terjadi dan bagaimana mereka dijalankan. Sekarang saat terjadi kesalahan, kita juga bisa memiliki sebuah catatan bagaimana dan kenapa bisa terjadi *bug*.
```
naturalisasi dan italic untuk english version

> Perhatikan semua action yang merubah state dari store diletakkan didalam store itu sendiri. Tipe pengelolaan state yang terpusat ini membuat semakin mudah untuk mengerti bahwa tipe perubahan apa yang bisa terjadi dan bagaimana mereka dijalankan. Sekarang saat terjadi kesalahan, kita juga memiliki sebuah catatan kenapa bisa terjadi bug.

Reviewer 2: ```suggestion
Perhatikan semua *action* yang merubah *state* diletakkan didalam store itu sendiri. Tipe pengelolaan state yang terpusat ini memudahkan kita untuk mengerti, tipe perubahan apa yang bisa terjadi dan bagaimana mereka dijalankan. Sekarang saat terjadi sebuah kesalahan, kita juga bisa memiliki sebuah catatan kenapa hal itu bisa menyebabkan munculnya *bug*.
```
inggris perlu italic, dan masih kurang natural

> Perhatikan semua *action* yang mengubah *state* dari store diletakkan didalam store itu sendiri. Tipe pengelolaan state yang terpusat ini membuatnya semakin mudah untuk dimengerti bahwa tipe perubahan apa saja yang bisa terjadi dan bagaimana mereka dijalankan. Sekarang saat terjadi kesalahan, kita juga memiliki sebuah catatan kenapa bisa terjadi bug.

Reviewer 2: store harus italic

> Sebagai tambahan, setiap instance/component tetap bisa memiliki dan mengelola state milik mereka sendiri:

Reviewer 2: ```suggestion
Sebagai tambahan, setiap *instance*/komponen tetap bisa memiliki dan mengelola state milik mereka sendiri:
```
italic untuk inggris

> Sebagai tambahan, setiap instance/component tetap bisa memiliki dan mengelola state milik mereka sendiri:

Reviewer 2: ```suggestion
Sebagai tambahan, setiap *instance*/komponen tetap bisa memiliki dan mengelola state milik mereka sendiri:
```

})
```

![Pengelolaan State](/images/state.png)

> <p class="tip">Sangat penting untuk mencatat bahwa kamu jangan pernah menggantikan state sebenarnya pada action mu - component dan store harus berbagi acuan objek yang sama supaya perubahan bisa diamati.</p>

Reviewer 2: ```suggestion
<p class="tip">Sebagai catatan penting, Anda jangan pernah menggantikan state asli pada *action* yang Anda buat - komponen dan *store* perlu untuk berbagi data referensi kedalam objek yang sama supaya perubahan bisa diamati.</p>
```
kamu => Anda
natural

> <p class="tip">Sangat penting untuk mencatat bahwa kamu jangan pernah menggantikan state sebenarnya pada action mu - component dan store harus berbagi acuan objek yang sama supaya perubahan bisa diamati.</p>

Reviewer 2: ```suggestion
<p class="tip">Sebagai catatan penting, jangan pernah mengganti *state* asli pada *actions* Anda - komponen dan *store* perlu untuk membagikan data referensi ke dalam objek yang sama supaya perubahan yang terjadi bisa diamati.</p>
```

> Saat kita melanjutkan pengembangan ketentuannya dimana component tidak boleh merubah state yang dimiliki oleh store secara langsung, tapi sebagai gantinya kita harus mengirim event untuk memberitahu store untuk menjalankan action, akhirnya kami bertemu dengan [Flux](https://facebook.github.io/flux/) arsiktektur. Keuntungan menggunakan ketentuan ini adalah kita bisa merekam semua perubahan state yang terjadi pada store dan sudah terimplement alat debugging seperti catatan perubahan, snapshot/potretan, dan bisa melakukan pengulangan sejarah/perjalanan waktu

Reviewer 3: Enaknya dikasih koma.
Saat kita melanjutkan pengembangan, ada ketentuan dimana komponen....

Reviewer 3: component -> komponen
merubah -> mengubah

Reviewer 3: terimplement -> terpasang

Reviewer 3: alat debugging -> alat debug

> Saat kita melanjutkan pengembangan ketentuannya dimana component tidak boleh merubah state yang dimiliki oleh store secara langsung, tapi sebagai gantinya kita harus mengirim event untuk memberitahu store untuk menjalankan action, akhirnya kami bertemu dengan [Flux](https://facebook.github.io/flux/) arsiktektur. Keuntungan menggunakan ketentuan ini adalah kita bisa merekam semua perubahan state yang terjadi pada store dan sudah terimplement alat debugging seperti catatan perubahan, snapshot/potretan, dan bisa melakukan pengulangan sejarah/perjalanan waktu

Reviewer 2: ```suggestion
Saat kita melanjutkan pengembangan, ketentuannya adalah komponen tidak boleh merubah *state* yang ada di dalam store secara langsung, tetapi kita harus mengirim sebuah *event* untuk memberitahu store bahwa dia harus menjalankan sebuah *action*, yang pada akhirnya kita akan merasakan proses seperti ini sama dengan dengan arsitektur milik [Flux](https://facebook.github.io/flux/). Keuntungan menggunakan aturan ini adalah kita bisa merekam semua perubahan *state* yang terjadi pada *store* dan bisa mengimplementasikan proses *debugging helpers* seperti catatan perubahan, *snapshot*, dan bisa melihat perubahan data secara historis/*time travel*.
```
Banyak yang perlu di natural kan

> Ini membawa kita kembali ke [vuex](https://github.com/vuejs/vuex), jadi jika kamu telah membacanya sejauh ini mungkin saatnya mencobanya!

Reviewer 2: ```suggestion
Mari kembali lagi ke [vuex](https://github.com/vuejs/vuex), jadi jika Anda telah membaca panduan pengelolaan *state* sampai sejauh ini, mungkin sekarang saatnya mencoba!
```
natural
